lib/logstash/codecs/csv.rb

```diff
@@ -51,113 +56,119 @@ class LogStash::Codecs::CSV < LogStash::Codecs::Base
   #     }
   config :convert, :validate => :hash, :default => {}
 
-  ##
-  # List of valid conversion types used for the convert option
-  ##
-  VALID_CONVERT_TYPES = [ "integer", "float", "date", "date_time", "boolean" ].freeze
-
-
   # The character encoding used in this codec. Examples include "UTF-8" and
   # "CP1252".
   config :charset, :validate => ::Encoding.name_list, :default => "UTF-8"
 
+  CONVERTERS = {
+    :integer => lambda do |value|
+      CSV::Converters[:integer].call(value)
+    end,
+
+    :float => lambda do |value|
+      CSV::Converters[:float].call(value)
+    end,
+
+    :date => lambda do |value|
+      result = CSV::Converters[:date].call(value)
+      result.is_a?(Date) ? LogStash::Timestamp.new(result.to_time) : result
+    end,
+
+    :date_time => lambda do |value|
+      result = CSV::Converters[:date_time].call(value)
+      result.is_a?(DateTime) ? LogStash::Timestamp.new(result.to_time) : result
+    end,
+
+    :boolean => lambda do |value|
+       value = value.strip.downcase
+       return false if value == "false"
+       return true  if value == "true"
+       return value
+    end
+  }
+  CONVERTERS.default = lambda {|v| v}
+  CONVERTERS.freeze
+
   def register
     @converter = LogStash::Util::Charset.new(@charset)
     @converter.logger = @logger
 
     # validate conversion types to be the valid ones.
-    @convert.each_pair do |column, type|
-      if !VALID_CONVERT_TYPES.include?(type)
-        raise LogStash::ConfigurationError, "#{type} is not a valid conversion type."
-      end
+    bad_types = @convert.values.select do |type|
+      !CONVERTERS.has_key?(type.to_sym)
+    end.uniq
+    raise(LogStash::ConfigurationError, "Invalid conversion types: #{bad_types.join(', ')}") unless bad_types.empty?
+
+    # @convert_symbols contains the symbolized types to avoid symbol conversion in the transform method
+    @convert_symbols = @convert.inject({}){|result, (k, v)| result[k] = v.to_sym; result}
```

Comment:
(totally a nitpick thread)

Are we constrained to Ruby 2.1+ (e.g., via running only on Logstash 6+?)

If so, this reads a bit clearer:

```suggestion
    @convert_symbols = @convert.map { | k, v| [k, v.to_sym] }.to_h
```

Reply:
Another possibly more clear option is `Enumerable#each_with_object`, which takes away the need for the accumulator to be returned from each iteration of the block.

```diff
-    @convert_symbols = @convert.inject({}){|result, (k, v)| result[k] = v.to_sym; result}
+    @convert_symbols = @convert.each_with_object({}){|(k, v), result| result[k] = v.to_sym}
```